index.bs

```diff
@@ -6219,6 +6200,48 @@ dictionary MLOperandDescriptor {
   </div>
 </details>
 
+Algorithms {#algorithms}
+=====================
+
+## Broadcasting ## {#algorithms-broadcasting}
+
+Broadcasting refers to how operations treat tensors with different shapes, and follow the precedent set by [[!numpy-broadcasting-rule]].
+
+<div algorithm>
+To <dfn data-lt="unidirectionally broadcasting the shapes">unidirectionally broadcast the shapes</dfn> |A| and |B|, perform the following steps. |A| and |B| are [=/lists=] of positive integers, representing the dimensions of tensors, and the steps return a new [=/list=] of positive integers, or failure.
+
+1. Let |sizeA| be the [=list/size=] of |A|.
+1. Let |sizeB| be the [=list/size=] of |B|.
+1. Let |output| be a new [=/list=].
+1. [=list/For each=] |index| in [=the range=] 0 to |sizeA|, exclusive:
+    1. Let |dimA| be |A|[|sizeA| - |index| - 1] if |index| < |sizeA|, or 1 otherwise.
+    1. Let |dimB| be |B|[|sizeB| - |index| - 1] if |index| < |sizeB|, or 1 otherwise.
+    1. If |dimA| is not equal to |dimB| and |dimA| is not equal to 1, then return failure.
+    1. [=list/Prepend=] |dimA| to |output|.
+1. Return |output|.
+
+</div>
+
+<div algorithm>
+|A| is <dfn>unidirectionally broadcastable</dfn> to |B| if [=unidirectionally broadcasting the shapes=] |A| and |B| does not result in failure.
+</div>
+
+<div algorithm>
+To <dfn data-lt="bidirectionally broadcasting the shapes">bidirectionally broadcast the shapes</dfn> |A| and |B|, perform the following steps. |A| and |B| are [=/lists=] of positive integers, representing the dimensions of tensors, and the steps return a new [=/list=] of positive integers, or failure.
+
+1. Let |sizeA| be the [=list/size=] of |A|.
+1. Let |sizeB| be the [=list/size=] of |B|.
+1. Let |outputSize| be the maximum of |sizeA| and |sizeB|.
+1. Let |output| be a new [=/list=].
+1. [=list/For each=] |index| in [=the range=] 0 to |outputSize|, exclusive:
+    1. Let |dimA| be |A|[|sizeA| - |index| - 1] if |index| < |sizeA|, or 1 otherwise.
```

Comment:
Another option that might be clearer and functionally equivalent, because then the loop becomes a more straight-forward foward pass using Append and without the negation, is to just pad A and B upfront (we do this often inside our internal code to avoid going crazy). e.g.

```
1. Let |outputSize| be the maximum of |sizeA| and |sizeB|.
1. Let |output| be a new [=/list=].
1. Let |paddedA| be a copy of |A| with enough leading 1's to pad up to length |outputSize|.
1. Let |paddedB| be a copy of |B| with enough leading 1's to pad up to length |outputSize|.
1. [=list/For each=] |index| in [=the range=] 0 to |outputSize|, exclusive:
    1. Let |dimA| be |paddedA|[|index|] if |index| < |sizeA|, or 1 otherwise.
    1. Let |dimB| be |paddedB|[|index|] if |index| < |sizeB|, or 1 otherwise.
    1. If |dimA| is not equal to |dimB|, and |dimA| is not equal to 1, and |dimB| is not equal to 1, then return failure.
    1. [=list/Append=] the maximum of |dimA| and |dimB| to |output|.
1. Return |output|.
```
🤷‍♂️ (no strong feeling either way)

Reply:
Yeah I get too clever sometimes. Suggestion SGTM